### PR TITLE
Adding basic API for memory profiling (src/training_hub/profiling)

### PR DESCRIPTION
**Pull request by mazam-lab**
The backend logic still needs some work on my end, but I'm opening the PR now to ensure the API signature and code standards is looking good. 

Please note that for 3.0, we'll only be adding a very basic memory profiling API. Hardware discovery, runtime estimations, hyperparameter and hardware recommendations, etc. are planned for 3.1+ 

## Summary by CodeRabbit

* **New Features**
  * Adds a memory estimation tool for fine‑tuning that reports per‑GPU VRAM needs (parameters, optimizer state, gradients, activations, outputs) and returns low/expected/high bounds.
  * Supports SFT and OSFT (including an experimental OSFT mode), Liger‑aware adjustments, and a unified top‑level estimate entry point.

* **Documentation**
  * Adds a notebook and README section demonstrating SFT/OSFT workflows, estimator options, and example outputs.

### CODE REVIEW

File: src/training_hub/profiling/memory_estimator.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                            'self_attn.o_proj',
+                            'mlp.gate_proj',
+                            'mlp.up_proj',
+                            'mlp.down_proj']
+    for term in TARGET_TERMS:
+        if layer_name.find(term) > -1 and layer_name.find('weight') > -1:
+            if layer_data.dim() < 2:
+                layer_data = layer_data.unsqueeze(0)
+            U_bytes_n: int = layer_data.shape[0] * layer_data.shape[0] 
+            S_bytes_n: int  = layer_data.shape[0] * layer_data.shape[1]
```
Comment (NikhilNayak-debug) [inline src/training_hub/profiling/memory_estimator.py:33]: Size of S should actually be the minimum of layer_data.shape[0] and layer_data.shape[1]. For scaling law analysis, memory footprint is dominated by other two parameters (U and V) since S only scales linearly (order 1). So S can be ignored for simplicity.

**Review by NikhilNayak-debug — COMMENTED** [review]
Overall this looks like a good draft!

File: src/training_hub/profiling/memory_estimator.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    if osft:
+        gpu_vram_par: int = _calc_osft_params(model) / num_gpus
+    else:
+        gpu_vram_par: int = num_params * FLOAT32_BYTES_N / num_gpus
+    gpu_vram_opt: int = FLOAT32_BYTES_N * num_trainable_params * ADAMW_PARAMS_N / num_gpus
+    gpu_vram_grad: int = FLOAT32_BYTES_N * num_trainable_params / num_gpus
+
+    # NOTE: Typically, K is somewhere in [10, 30], but some research suggests that K=17.
+    # TODO: Add papers
+    gpu_vram_act_low: int = tokens_per_gpu * FLOAT32_BYTES_N * 10 * num_layers * hidden_size
```
[the remark this answers is not in the document]
Comment (mazam-lab) [reply src/training_hub/profiling/memory_estimator.py:125]: Very shortly after I did this PR I made some discoveries that change this formula:

Basically, the activations are made up of two components:
1. The mid-layer activations (gpu_vram_act), which can be calculated as max_tokens_per_gpu * FLOAT32_BYTES_N * num_layers * hidden_size (set K to 1)
2. The output tensor (gpu_vram_out), which is a tensor of size max_tokens_per_gpu x vocab_size x FLOAT32_BYTES_N. 

From what I've seen empirically, you can get a reasonably good estimate if you just use gpu_vram_act + (gpu_vram_out * 2) as your activation memory estimate

With liger kernels enabled you can simply use gpu_vram_act and ignore gpu_vram_out.

But the theoretical justification on using these two formulas is still a bit wishy-washy, so I want to tighten that up a bit if possible

File: src/training_hub/profiling/memory_estimator.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    gpu_vram_act_mid: int = tokens_per_gpu * FLOAT32_BYTES_N * 17 * num_layers * hidden_size
+
+    # Lambda functions to calculate the total amount of VRAM and misc overhead
+    calc_subtotal: Callable[[int], int] = \
+        lambda act_value: gpu_vram_par + gpu_vram_opt + gpu_vram_grad + act_value
+    calc_overhead: Callable[[int, int], int] = \
+        lambda tot_val, act_val: tot_val - gpu_vram_par - gpu_vram_opt - gpu_vram_grad - act_val
+
+    # Perform those calculations using the lambdas!
+    gpu_vram_total_high: int = int(1.3 * calc_subtotal(gpu_vram_act_high))
```
[the remark this answers is not in the document]
Comment (mazam-lab) [reply src/training_hub/profiling/memory_estimator.py:136]: Most formulas for calculating memory add about 10-30% to the initial subtotal to account for overhead, hence the scalars used here

File: src/training_hub/profiling/memory_estimator.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    given hyperparameters. After that, determine how possible it is for
+    the given hardware to run the model, and note how much more memory
+    is needed to make the training more feasible. 
+
+    Note that this estimate assumes training_hub will be used, 
+    in which all data types are float32 and the optimizer is always AdamW.
+
+    Args:
+        num_gpus: Number of GPUs to use for training (default: 8)
+        gpu_memory: The VRAM of each GPU in bytes (default: 24000000000)
```
Comment (RobotSail) [inline src/training_hub/profiling/memory_estimator.py:74]: Readability 

```suggestion
        gpu_memory: The VRAM of each GPU in bytes (default: 24GB)
```

File: src/training_hub/profiling/memory_estimator.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        model_path: Path to the model to fine-tune, namely a HuggingFace
+                    model path (default: "Qwen/Qwen2.5-7B-Instruct")
+        effective_batch_size: Effective batch size for training, typically
+                              not as important as max_tokens_per_gpu
```
Comment (RobotSail) [inline src/training_hub/profiling/memory_estimator.py:78]:
```suggestion
        effective_batch_size: The number of samples in a minibatch that the model has to see before backpropping.
```

File: src/training_hub/profiling/memory_estimator.py (hunk 1 above)
Comment (mazam-lab) [reply src/training_hub/profiling/memory_estimator.py:40]: Oh, so S can be reduced in size since it's a diagonal matrix? Makes sense, did.

File: src/training_hub/profiling/memory_estimator.py (hunk 4 above)
Comment (mazam-lab) [reply src/training_hub/profiling/memory_estimator.py:82]: did, also adjusted the defaults to be reflected in the code and changed the default gpu_memory value to 85899345920 (for 80 GB)

File: src/training_hub/profiling/memory_estimator.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    osft: bool = False
+    ) -> tuple[int, int, int]:
+
+    """
+    Calculate the memory needed to fine tune the given model for the 
+    given hyperparameters. After that, determine how possible it is for
+    the given hardware to run the model, and note how much more memory
+    is needed to make the training more feasible. 
+
+    Note that this estimate assumes training_hub will be used,
```
[the remark this answers is not in the document]
Comment (mazam-lab) [reply src/training_hub/profiling/memory_estimator.py:77]: Okay, so the point of this comment is to let the user know that we'll be assuming all tensors placed on VRAM will be Float32; so should we still clarify that the underlying backend calculations might not necessarily be entirely done in Float32?

**Comment by mazam-lab** [comment]
> Thank you for the PR @mazam-lab, I took a look and left a few comments. Since the calculator is implemented as a script, maybe you could also add a notebook that provides a few examples on its usage?

Yeah I can do that; should the notebook provide some empirical results on what the actual memory usage for some select models and training setups are? (i.e. have a table showing the actual memory usage of Granite 2B on OSFT with Liger Kernels and 16,384 tokens)

File: src/training_hub/profiling/memory_estimator.py [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+FLOAT8_BYTES_N: int = 1
+ADAMW_PARAMS_N: int = 2
+
+# Helper function to do the rounding when printing 
+def ROUNDER(value: int) -> str: return str(round(value / 1073741824, 1))
+
+# Helper function to calculate how much the given unfrozen_rank_ratio 
+# will be affecting the OSFT estimation (through a quadratic mapping where
+# 0 is 0.25 of SFT's value, 0.33 is equal to SFT's value, and 1 is twice of SFT's value)
+def OSFT_RATIO(value: float) -> float: return -0.7802 * (value) * (value) + 2.5302 * (value) + 0.25
```
[the remark this answers is not in the document]
Comment (mazam-lab) [reply src/training_hub/profiling/memory_estimator.py]: sure, that could work too. if you're confident in that estimation I'll go ahead and implement it
